File: examples/android/app/src/main/java/com/jnibridge/modules/UIModule.java
package com.jnibridge.modules;

import android.util.Log;

import com.jnibridge.EngineScope;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;

public class UIModule extends ModuleBase {
    public UIModule(EngineScope scope) {
        super("UI", scope);
    }

    @EngineMethod(name = "createElement")
    public String createElement(int callId, String args) {
        try {
            Log.d("JSEngine", "createElement: " + args);
            JSONObject json = new JSONObject(args);
            String tagName = json.getString("tagName");
            int uniqueId = json.getInt("uniqueId");
            getScope().getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    getScope().getUIRender().createElement(uniqueId, tagName);
                }
            });
        } catch (JSONException e) {
            e.printStackTrace();
            Log.d("JSEngine", "createElement: " + args);
        }
        return "";
    }

    @EngineMethod(name = "setStyle")
    public String setStyle(int callId, String args) {
        Log.d("JSEngine", "setStyle: " + args);
        try {
            JSONObject json = new JSONObject(args);
            int uniqueId = json.getInt("uniqueId");
            boolean replace = json.has("replace") && json.getInt("replace") == 1;
            if(replace) {
                getScope().getUIRender().resetStyle(uniqueId);
            }
            JSONObject style = json.getJSONObject("style");
            Iterator<String> keys = style.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                String value = style.getString(key);
                getScope().getUIRender().setStyle(uniqueId, key, value);
            }
        } catch (JSONException jsonException) {
            jsonException.printStackTrace();
        }
        return "";
    }

    @EngineMethod(name = "setAttribute")
    public String setAttribute(int callId, String args) {
        Log.d("JSEngine", "setAttribute: " + args);
        try {
            JSONObject json = new JSONObject(args);
            int uniqueId = json.getInt("uniqueId");
            JSONObject attribute = json.getJSONObject("attribute");
            Iterator<String> keys = attribute.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                String value = attribute.getString(key);
                getScope().getUIRender().setAttribute(uniqueId, key, value);
            }
        } catch (JSONException jsonException) {
            jsonException.printStackTrace();
        }
        return "";
    }

    @EngineMethod(name = "removeAttribute")
    public String removeAttribute(int callId, String args) {
        Log.d("JSEngine", "setAttribute: " + args);
        try {
            JSONObject json = new JSONObject(args);
            int uniqueId = json.getInt("uniqueId");
            String key = json.getString("key");
            getScope().getUIRender().removeAttribute(uniqueId, key);
        } catch (JSONException jsonException) {
            jsonException.printStackTrace();
        }
        return "";
    }

    @EngineMethod(name = "getAttribute")
    public String getAttribute(int callId, String args) {
        Log.d("JSEngine", "getAttribute: " + args);
        try {
            JSONObject json = new JSONObject(args);
            int uniqueId = json.getInt("uniqueId");
            String key = json.getString("key");
            return getScope().getUIRender().getAttribute(uniqueId, key);
        } catch (JSONException jsonException) {
            jsonException.printStackTrace();
        }
        return "";
    }

    @EngineMethod(name = "insertChild")
    public String insertChild(int callId, String args) {
        Log.d("JSEngine", "insertChild: " + args);
        try {
            JSONObject json = new JSONObject(args);
            int uniqueId = json.getInt("uniqueId");
            JSONObject newChild = json.getJSONObject("newChild");
            int newChildId = newChild.getInt("uniqueId");
            JSONObject refChild = json.getJSONObject("refChild");
            int refChildId = refChild.getInt("uniqueId");
            getScope().getUIRender().insertChild(uniqueId, newChildId, refChildId);
        } catch (JSONException jsonException) {
            jsonException.printStackTrace();
        }
        return "";
    }

    @EngineMethod(name = "replaceChild")
    public String replaceChild(int callId, String args) {
        Log.d("JSEngine", "replaceChild: " + args);
        try {
            JSONObject json = new JSONObject(args);
            int uniqueId = json.getInt("uniqueId");
            JSONObject newChild = json.getJSONObject("newChild");
            int newChildId = newChild.getInt("uniqueId");
            JSONObject oldChild = json.getJSONObject("oldChild");
            int oldChildId = oldChild.getInt("uniqueId");
            getScope().getUIRender().replaceChild(uniqueId, newChildId, oldChildId);
        } catch (JSONException jsonException) {
            jsonException.printStackTrace();
        }
        return "";
    }

    @EngineMethod(name = "scrollTo")
    public String scrollTo(int callId, String args) {
        Log.d("JSEngine", "scrollTo: " + args);
        try {
            JSONObject json = new JSONObject(args);
            int uniqueId = json.getInt("uniqueId");
            float x = (float)json.getDouble("x");
            float y = (float)json.getDouble("y");
            getScope().getUIRender().scrollTo(uniqueId, x, y);
        } catch (JSONException jsonException) {
            jsonException.printStackTrace();
        }
        return "";
    }
}

File: examples/ios/ioscplus/ioscplus/UIModule.swift
//
//  JSTimerModule.swift
//  ioscplus
//
//  Created by song on 2021/8/5.
//

import Foundation



public struct CElementArgs: Decodable {
    let uniqueId: Int32;
    let tagName: String?;
}

public class UIModule : NativeModule {
    public init(scope: JSEngineScope) {
        super.init(moduleName: "UI", scope: scope)
    }

    public override func invokeMethod(callId: Int32, methodName: String, args: String) -> String {
        var result = ""
        switch methodName {
            case "createElement":
                createElement(args: args)
                break;
            case "setStyle":
                setStyle(args: args)
                break;
            case "setAttribute":
                setAttribute(args: args)
                break;
            case "removeAttribute":
                removeAttribute(args: args)
                break;
            case "getAttribute":
                result = getAttribute(args: args)
                break;
            case "insertChild":
                insertChild(args: args)
                break;
            case "replaceChild":
                replaceChild(args: args)
                break;
            default:
                break;
        }
        return result;
    }
    
    
    private func createElement(args: String) {
        if let data = args.data(using: String.Encoding.utf8, allowLossyConversion: true) {
            if let createElementArgs = try? JSONDecoder().decode(CElementArgs.self, from: data) {
                getScope().getUIRender().createElement(uniqueId: createElementArgs.uniqueId, tagName: createElementArgs.tagName!)
            }
        }
    }
    
    private func setStyle(args: String) {
        if(!getScope().getUIRender().isReady()) {
            return;
        }
        if let data = args.data(using: String.Encoding.utf8, allowLossyConversion: true) {
            if let setStyleArgs = try? JSONDecoder().decode(SetStyleArgs.self, from: data) {
                if(setStyleArgs.replace == 1) {
                    getScope().getUIRender().resetStyle(uniqueId: setStyleArgs.uniqueId)
                }
                setStyleArgs.style.forEach { (key: String, value: String) in
                    getScope().getUIRender().setStyle(uniqueId: setStyleArgs.uniqueId, key: key, value: value)
                }
            }
        }
    }

    public func setAttribute(args: String) {
        if let data = args.data(using: String.Encoding.utf8, allowLossyConversion: true) {
            if let setAttributeArgs = try? JSONDecoder().decode(SetAttributeArgs.self, from: data) {
                setAttributeArgs.attribute.forEach { (key: String, value: String) in
                    getScope().getUIRender().setAttribute(uniqueId: setAttributeArgs.uniqueId, key: key, value: value)
                }
            }
        }
    }

    public func removeAttribute(args: String) {
        if let data = args.data(using: String.Encoding.utf8, allowLossyConversion: true) {
            if let removeAttributeArgs = try? JSONDecoder().decode(RemoveAttributeArgs.self, from: data) {
                getScope().getUIRender().removeAttribute(uniqueId: removeAttributeArgs.uniqueId, key: removeAttributeArgs.key)
            }
        }
    }

    public func getAttribute(args: String) -> String {
        if let data = args.data(using: String.Encoding.utf8, allowLossyConversion: true) {
            if let getAttributeArgs = try? JSONDecoder().decode(GetAttributeArgs.self, from: data) {
                return getScope().getUIRender().getAttribute(uniqueId: getAttributeArgs.uniqueId, key: getAttributeArgs.key)
            }
        }
        return ""
    }

    public func insertChild(args: String) {
        if let data = args.data(using: String.Encoding.utf8, allowLossyConversion: true) {
            if let insertChildArgs = try? JSONDecoder().decode(InsertChildArgs.self, from: data) {
                getScope().getUIRender().insertChild(uniqueId: insertChildArgs.uniqueId, newChildId: insertChildArgs.newChild.uniqueId, refChildId: insertChildArgs.refChild.uniqueId)
            }
        }
    }

    public func replaceChild(args: String) {
        if let data = args.data(using: String.Encoding.utf8, allowLossyConversion: true) {
            if let replaceChildArgs = try? JSONDecoder().decode(ReplaceChildArgs.self, from: data) {
                getScope().getUIRender().replaceChild(uniqueId: replaceChildArgs.uniqueId, newChildId: replaceChildArgs.newChild.uniqueId, oldChildId: replaceChildArgs.oldChild.uniqueId)
            }
        }
    }

    public func scrollTo(args: String) {
        if let data = args.data(using: String.Encoding.utf8, allowLossyConversion: true) {
            if let scrollToArgs = try? JSONDecoder().decode(ScrollToArgs.self, from: data) {
                getScope().getUIRender().scrollTo(uniqueId: scrollToArgs.uniqueId, x: scrollToArgs.x, y: scrollToArgs.y)
            }
        }
    }
}
